tuf/data/keys.go

```diff
@@ -34,38 +28,177 @@ type KeyPair struct {
 	Private []byte `json:"private"`
 }
 
-// TUFKey is the structure used for both public and private keys in TUF.
-// Normally it would make sense to use a different structures for public and
-// private keys, but that would change the key ID algorithm (since the canonical
-// JSON would be different). This structure should normally be accessed through
-// the PublicKey or PrivateKey interfaces.
-type TUFKey struct {
-	id    string
-	Type  KeyAlgorithm `json:"keytype"`
-	Value KeyPair      `json:"keyval"`
+// Keys represents a map of key ID to PublicKey object. It's necessary
+// to allow us to unmarshal into an interface via the json.Unmarshaller
+// interface
+type Keys map[string]PublicKey
+
+// UnmarshalJSON implements the json.Unmarshaller interface
+func (ks *Keys) UnmarshalJSON(data []byte) error {
+	parsed := make(map[string]tufKey)
+	err := json.Unmarshal(data, &parsed)
+	if err != nil {
+		return err
+	}
+	final := make(map[string]PublicKey)
+	for k, tk := range parsed {
+		final[k] = typedPublicKey(tk)
+	}
+	*ks = final
+	return nil
+}
+
+// KeyList represents a list of keys
+type KeyList []PublicKey
+
+// UnmarshalJSON implements the json.Unmarshaller interface
+func (ks *KeyList) UnmarshalJSON(data []byte) error {
+	parsed := make([]tufKey, 0, 1)
+	err := json.Unmarshal(data, &parsed)
+	if err != nil {
+		return err
+	}
+	final := make([]PublicKey, 0, len(parsed))
+	for _, tk := range parsed {
+		final = append(final, typedPublicKey(tk))
+	}
+	*ks = final
+	return nil
+}
+
+func typedPublicKey(tk tufKey) PublicKey {
+	switch tk.Algorithm() {
+	case ECDSAKey:
+		return &ECDSAPublicKey{tufKey: tk}
+	case ECDSAx509Key:
+		return &ECDSAx509PublicKey{tufKey: tk}
+	case RSAKey:
+		return &RSAPublicKey{tufKey: tk}
+	case RSAx509Key:
+		return &RSAx509PublicKey{tufKey: tk}
+	case ED25519Key:
+		return &ED25519PublicKey{tufKey: tk}
+	}
+	return &UnknownPublicKey{tufKey: tk}
```

Comment:
In the context of TUF, it's incorrect to error on a key or signature type you don't recognize. You simply won't be able to use it for signing/validating signatures, but it doesn't mean the data is invalid. We already handled this case in the same manner, it was just less explicit and hidden under the fact everything was a generic `TUFKey` or `Signature` object.